### PR TITLE
feat: poseidon builtin runner

### LINKED ISSUE #314: feat: poseidon builtin runner
implement poseidon builtin runner

[Rust implementation](https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/vm/runners/builtin_runner/poseidon.rs)

### PR DESCRIPTION
**Pull request by StringNick**
closes #314

### CODE REVIEW

File: src/vm/builtins/builtin_runner/poseidon.zig [hunk 1]
```diff
@@ -54,37 +69,362 @@ pub const PoseidonBuiltinRunner = struct {
...
+            @intCast(self.n_input_cells),
+        )) {
+            return null;
+        }
+
+        const felt = self.cache.get(address);
+
+        if (felt != null) {
+            return .{ .felt = felt.? };
+        }
```
Comment (tcoratger) [inline src/vm/builtins/builtin_runner/poseidon.zig:194]:
```zig
if (self.cache.get(address)) |felt| return .{ .felt = felt };
```

File: src/vm/builtins/builtin_runner/poseidon.zig [hunk 2]
```diff
@@ -54,37 +69,362 @@ pub const PoseidonBuiltinRunner = struct {
...
+                @intCast(address.offset),
+            ),
+            @as(
+                usize,
+                @intCast(self.cells_per_instance),
+            ),
+        );
+        if (index < @as(
+            usize,
+            @intCast(self.n_input_cells),
```
Comment (tcoratger) [inline src/vm/builtins/builtin_runner/poseidon.zig:185]: Here as `n_input_cells` is u32 and you want to cast to usize, I don't think that you have to write that, type coercion should be automatic and implicit with

```zig
if (index < self.n_input_cells)
```

File: src/vm/builtins/builtin_runner/poseidon.zig [hunk 3]
```diff
@@ -54,37 +69,362 @@ pub const PoseidonBuiltinRunner = struct {
...
     pub fn initialStack(self: *Self, allocator: Allocator) !ArrayList(MaybeRelocatable) {
-        _ = self;
         var result = ArrayList(MaybeRelocatable).init(allocator);
-        errdefer result.deinit();        
+        if (self.included) {
+            try result.append(MaybeRelocatable.fromSegment(
+                @intCast(self.base),
+                0,
+            ));
+            return result;
```
Comment (tcoratger) [inline src/vm/builtins/builtin_runner/poseidon.zig:89]: You can remove this return as long as you also return `result` after your if statement no?

**Review by jobez — COMMENTED** [review]
one comment about a doc string, as well as you are missing an impl of `getMemorySegmentAddresses`

please follow this diff for conventions on how to line it up to the cairo_runner: https://github.com/keep-starknet-strange/ziggy-starkdust/pull/324/files

**Comment by StringNick** [comment]
@jobez  done

**Comment by jobez** [comment]
@StringNick lets resolve conflicts here and should be good to merge!

**Comment by StringNick** [comment]
> @StringNick lets resolve conflicts here and should be good to merge!

done

File: src/vm/builtins/builtin_runner/builtin_runner.zig [hunk 4]
```diff
@@ -126,7 +126,7 @@ pub const BuiltinRunner = union(enum) {
             .RangeCheck => |range_check| range_check.deduceMemoryCell(address, memory),
             .Keccak => |*keccak| try keccak.deduceMemoryCell(allocator, address, memory),
             .Signature => |signature| signature.deduceMemoryCell(address, memory),
-            .Poseidon => |poseidon| poseidon.deduceMemoryCell(address, memory),
+            .Poseidon => |*poseidon| try poseidon.deduceMemoryCell(allocator, address, memory),
```
Comment (jobez) [inline src/vm/builtins/builtin_runner/builtin_runner.zig:129]: one more ask: see in this change where we added a `getMemorySegmentAddresses` for the builtin_runner

we want to add this completed poseidons impl of `getMemorySegmentAddresses` there, then lgtm!

**Review by jobez — CHANGES_REQUESTED** [review]

**Review by jobez — APPROVED** [review]
lgtm

**Pull request merged**